compiler/codegen/OMRPeephole.cpp

```diff
@@ -0,0 +1,80 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "codegen/Peephole.hpp"
+#include "codegen/CodeGenerator.hpp"
+#include "codegen/Instruction.hpp"
+
+OMR::Peephole::Peephole(TR::Compilation* comp) :
+   _comp(comp), _cg(comp->cg())
+   {}
+
+TR::Peephole*
+OMR::Peephole::self()
+   {
+   return static_cast<TR::Peephole*>(this);
+   }
+
+TR::Compilation*
+OMR::Peephole::comp() const
+   {
+   return _comp;
+   }
+
+TR::CodeGenerator*
+OMR::Peephole::cg() const
+   {
+   return _cg;
+   }
+
+bool
+OMR::Peephole::perform()
+   {
+   bool performed = false;
+
+   TR::Instruction* prevInst = NULL;
+   TR::Instruction* currInst = self()->cg()->getFirstInstruction();
+
+   while (currInst != NULL)
+      {
+      performed |= self()->performOnInstruction(currInst);
+
+      // The current instruction being processed may have been removed, moved, or altered in some way by the peephole
+      // optimization. In such cases we need to ensure we resume processing instructions from a valid point.
+      if (currInst->getPrev() == prevInst)
+         {
+         prevInst = currInst;
+         currInst = currInst->getNext();
+         }
+      else
+         {
+         currInst = prevInst->getNext();
+         }
```

Comment:
> Do we need to document a set of assumptions that peepholes must obey for this code to work?

I think we should make our life simple and just document the restriction. Peepholes should _only_ modify instructions following (or at) the current instruction being examined, never before. This is currently true for all peepholes that exist in the codebase today.

Reply:
Fixed in b73252e.

Reply:
I'm not sure this is actually true. `recordFormPeephole` in particular will walk backwards after finding a `cmpdi cr0, rX, 0` instruction to try and find the instruction that wrote to rX and might change that instruction to make it record-form. In fact, for `rlwinm` instructions on older versions of Power, it might remove it entirely and replace it with an `andi.` or `andis.` instruction.

Reply:
You're right. The piece of code in question is in [1], and there is another example just a few lines before. I've searched for all references to `->getPrev` in the Power peepholes and this is the only place where we could potentially remove the previous instruction as well as the current instruction. In such a case the code being reviewed here is problematic as the previous instruction will have been unlinked from the instruction stream.

How to fix this?

1. Expose `prevInst` and `currInst` as fields rather than local variables in the above algorithm, which extensions can modify as they see fit
    - In the Power case we would modify these pointers when we are about to perform the record form peephole so it knows where to restart
2. We augment all peepholes to return the restart point
3. We backpedal on f4d073637e6e751673c0cbd32f98c7dd66cd8b2c and somehow modify this algorithm to handle the pointers not being set to `NULL`
4. ...?

I'm in favor of 1. to handle these cases. I don't know if there are other alternatives that people can think of. 2. seems quite invasive as every peephole would suffer just so we can accommodate this one exceptional case. 3. seems opposite of where we want to be. Why should dangling instructions exist? @aviansie-ben thoughts on the options or alternatives?

[1] https://github.com/eclipse/omr/pull/5235/files#diff-bdc496687bbdf4ac8bd615034145e891R201-R207

Reply:
Thinking about this some more, option 4. is possible if we do something like:

```
   while (currInst != NULL)
      {
      performed |= self()->performOnInstruction(currInst);

      // The current instruction being processed may have been removed, moved, or altered in some way by the peephole
      // optimization. In such cases we need to ensure we resume processing instructions from a valid point.
      if (currInst->getPrev()->getNext() == currInst)
         {
         prevInst = currInst;
         currInst = currInst->getNext();
         }
      else
         {
         currInst = currInst->getPrev()->getNext();
         }
      }
```

which may be more appealing. This works even in the case we remove both `currInst` and `prevInst`, because the removal has to happen one instruction at a time. @aviansie-ben do you think we should prefer option 4. in light of this?

Reply:
I'd be alright with exposing `prevInst` as a field to get this PR moved along. In the long term, I'd prefer that peepholes specifically return their restart points (possibly by modifying the `cursor` field?) since having two instruction cursors as fields seems like it could cause confusion IMO. That being said, I don't think this is an issue worth delaying this PR over, since it would involve more invasive changes to the peepholes themselves.

Reply:
Well, it looks like we once again ended up posting comments at nearly exactly the same time... I think the solution you've come up with regarding option 3 might work given the current set of peepholes, but it doesn't work in the general case.

Imagine that we start with a sequence of instructions `A, B, C, D, E` and start a peephole at `C`, so `prevInst = B` and `currInst = C`. We start by removing `C`, leaving `C->getPrev() = B`. Next, we remove `B`, leaving `B->getNext() = D`. From there, we can end up in a problematic state by removing `D`. In this case `C->getPrev()->getNext() = D`, so we end up restarting peepholes starting from the removed instruction `D`. Granted, this is a pretty contrived series of events, but it could still theoretically occur.

Reply:
> Granted, this is a pretty contrived series of events, but it could still theoretically occur.

I agree. Although very unlikely it is still possible and that is deterrent enough. I'm going to go ahead with implementing 1. in a separate commit so it's easy to review. In the future I also agree we need a proper way for peepholes to indicate a restart point. However we are not there just yet.

Reply:
@aviansie-ben I've gone ahead and implemented the fix in 79d2fe3. I'm not happy with the solution but it seems to be the least invasive one which doesn't interfere with other peepholes and allows us to proceed with the work here.

I've summarized the problem in the commit message for future reference. If you are ok with this fix I will go ahead and open up an issue and write down some thoughts so we can remove this hack and think about peephole traversals, peephole windows, restart points, etc. Let me know!